### ISSUE TITLE
mocktail_image_network: `mockNetworkImages` affects other tests

### ISSUE THREAD

**Issue by jjochen** [felangel/mocktail#134]
**Describe the bug**

Mocking network images in one test will mock them for all upcoming tests in the same file. 

**To Reproduce**
Updated tests from the [example](https://github.com/felangel/mocktail/blob/2fbf56235eff60596e79bf2031fe639e38f769d4/packages/mocktail_image_network/example/test/image_test.dart):
```

void main() {
  testWidgets('can use mocktail for network images', (tester) async {
    await mockNetworkImages(() async {
      await tester.pumpWidget(FakeApp());
      expect(find.byType(Image), findsOneWidget);
    });
  });

  testWidgets('this test should fail', (tester) async {
    // This test fails if run separately without the above test (as expected),
    // but succeeds when run after the above test.

    await tester.pumpWidget(FakeApp());
  });
}
```
Running the second test after the first succeeds unexpectedly.

**Expected behavior**
Mocking network images in one test should not affect other tests.

**Comment by kturney** [comment]
FWIW, we started experiencing a similar issue in our local fork after down streaming https://github.com/felangel/mocktail/pull/120.

Turns out the the first mock ended up in the [`imageCache`](https://github.com/flutter/flutter/blob/4be65314d585961cabb5dcab1ea36d0aa5e18362/packages/flutter/lib/src/painting/image_provider.dart#L518), and so future image requests never called [`_createHttpClient`](https://github.com/felangel/mocktail/blob/main/packages/mocktail_image_network/lib/src/mocktail_image_network.dart#LL65C12-L65C29).

We fixed it by implementing our local `mockNetworkImages` like so

```dart
Future<T> mockNetworkImages<T>(FutureOr<T> Function() body) async {
  final currentClientProvider = debugNetworkImageHttpClientProvider;
  try {
    debugNetworkImageHttpClientProvider = _createHttpClient;
    return await body();
  } finally {
    debugNetworkImageHttpClientProvider = currentClientProvider;
    TestWidgetsFlutterBinding.ensureInitialized().imageCache.clear();
  }
}
```

**Comment by felangel** [comment]
@jjochen sorry for the delay! Can you verify whether the following resolves the issue?

```dart
void main() {
  testWidgets('can use mocktail for network images', (tester) async {
    final app = mockNetworkImages(() => FakeApp());
    await tester.pumpWidget(app);
    expect(find.byType(Image), findsOneWidget);
  });

  testWidgets('this test should fail', (tester) async {
    await tester.pumpWidget(FakeApp());
  });
}
```

@kturney are you able to provide a minimal reproduction sample?

**Comment by kturney** [comment]
Unfortunately, I probably won't have a chance to MR a failing test anytime soon.
But I can describe the failure we were seeing.

We had a test that runs with several variants.
The test expects has an expect that checks for a loading state, expecting the image to not yet be loaded.
The expect would pass for the first variant, but all subsequent variants would fail because the image was already loaded and so skipped past showing loading state.

I hypothesize we didn't encounter this prior to https://github.com/felangel/mocktail/pull/120 because the request was never added to the imageCache since we hadn't been telling the imageCache the request was done.

But now subsequent tests just immediately load the image from imageCache.

**Comment by felangel** [comment]
I'm wondering whether scoping the `mockNetworkImage` to exclusively the widget would also resolve the issues you were encountering.

BAD:

```dart
void main() {
  testWidgets('...', (tester) async {
    await mockNetworkImages(() async {
      await tester.pumpWidget(MyWidget());
      // ...
    });
  });
}
```

GOOD:

```dart
void main() {
  testWidgets('...', (tester) async {
    final widget = mockNetworkImages(() => MyWidget());
    await tester.pumpWidget(widget);
    // ...
  });
}
```

**Comment by kturney** [comment]
At least for our use case, we enforce only using the "BAD" form because otherwise we encountered spurious image requests that occurred outside the mock and would therefore get a `400`.

We also use `debugNetworkImageHttpClientProvider` instead of `HttpOverrides` to work around https://github.com/flutter/flutter/issues/83901.

I see there was an apparently abandoned attempt to fix `HttpOverides` in https://github.com/flutter/flutter/pull/83952.

**Comment by SimonWeidemann** [comment]
@felangel Can we adapt the examples to your scoping solution and mention that in the description?
So that other people are aware that incorrect scoping leads to problems?
I am willing to create a pull request for this.

**Comment by felangel** [comment]
> @felangel Can we adapt the examples to your scoping solution and mention that in the description?
> So that other people are aware that incorrect scoping leads to problems?
> I am willing to create a pull request for this.

Yes definitely! PRs are welcome otherwise I should have time later today.

**Comment by felangel** [comment]
@jjochen you'll need to also move the `pumpWidget` call into the `mockNetworkImages closure.

Since this issue is quite old, I'm going to close but if this is still an issue for folks let me know (and ideally provide a minimal reproduction sample) and I'm happy to take a closer look at what we can do to improve things, thanks!

**Comment by ryzizub** [comment]
@felangel I still find this an issue. I'm using `mockNetworkImages` and also in same group of tests im doing two tests that are cowering errorBuilder, so im mock trough HttpOverride failing HttpClient. Seems that if I run those tests separately it's running ok, but if I run them as a group those which are running later are failing cause either there is mocked failing client or successful client even that in each test is should have own zone

**Comment by felangel** [comment]
> @felangel I still find this an issue. I'm using `mockNetworkImages` and also in same group of tests im doing two tests that are cowering errorBuilder, so im mocking trough HttpOverride failing HttpClient. Seems that if I run those tests separately it's running ok, but if I run them as a group those which are running later are failing cause either there is mocked failing client or successful client even that in each test is should have own zone

Can you file a new issue with a link to a minimal reproduction sample? Thanks! 🙏

**Comment by ryzizub** [comment]
@felangel here it is https://github.com/felangel/mocktail/issues/250 . Thanks!

**Comment by felangel** [comment]
This should be fixed in v1.3.0 🎉
